[Core 6.2] TTBR messages will no longer have DLQ flag by default

andreasohlund opened the pull request:
Fixes #2433 

### WIP

* Needs core 6.2 to be released

andreasohlund commented:
@SimonCropp @timbussmann do we need to mention this in the upgrade guide? (on the fence since this is really a bug fix)

SimonCropp commented:
> do we need to mention this in the upgrade guide

i dont think so

andreasohlund commented:
This one is ready for review

SimonCropp commented on nservicebus/msmq/dead-letter-queues.md:
```diff
@@ -0,0 +1,22 @@
+---
```
should this be added to the menu?

andreasohlund commented:
Menu entry added

On Tue, Feb 14, 2017 at 12:02 PM, Simon Cropp <notifications@github.com>
wrote:

> *@SimonCropp* commented on this pull request.
> ------------------------------
>
> In nservicebus/msmq/dead-letter-queues.md
> <https://github.com/Particular/docs.particular.net/pull/2437#pullrequestreview-21725863>
> :
>
> > @@ -0,0 +1,22 @@
> +---
>
> should this be added to the menu?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Particular/docs.particular.net/pull/2437#pullrequestreview-21725863>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHoZPjei6gHTc7-m-uqTM9TUsp7yKs_ks5rcYm7gaJpZM4L_nYH>
> .
>

indualagarsamy commented on nservicebus/msmq/dead-letter-queues.md:
```diff
@@ -0,0 +1,22 @@
+ - Transport
+---
+
+[Dead Letter Queues](https://msdn.microsoft.com/en-us/library/ms706227.aspx) is a feature of MSMQ that tracks messages that are undeliverable, deleted, expired etc. NServiceBus endpoints will by default enable DLQ for all outgoing messages except messages that have a [Time To Be Received(TTBR)](/nservicebus/messaging/discard-old-messages.md) set. This avoids expired messages ending up in the DLQ  ...
```
@andreasohlund - Since there is DLQ and also TDLQ in the system queues, we should be more specific that we are talking about the T-DLQ.  Maybe add something to the effect, since in NServiceBus we are only dealing with transactional queues, we are more interested in T-DLQ or something to that effect first?

andreasohlund replied:
You can use non transactional queues by setting `useTransactionalQueues` in the [connection string](https://docs.particular.net/nservicebus/msmq/connection-strings)?

indualagarsamy replied:
How about:

@andreasohlund - Maybe we should a little more info. For example:

Instead of 
>This avoids expired messages ending up in the DLQ and wasting disk space on the machine where they timed out.

how about:

This avoids expired messages ending up either in the TDLQ or DLQ (if [non-transactional queues are used](/nservicebus/msmq/connection-strings.md)) wasting disk space on the machine when they timed out.

indualagarsamy commented on nservicebus/msmq/dead-letter-queues.md:
```diff
@@ -0,0 +1,22 @@
+
+### Enabling DLQ for messages with TTBR
+
+In Versions below 6.2 DLQ will be enabled by default for TTBR messages as well.
```
>In Versions below 6.2

Should be In Versions 6.2 and below, 

See the version language here:
https://github.com/particular/docs.particular.net/#version-language

andreasohlund replied:
Fixed

indualagarsamy commented on nservicebus/msmq/dead-letter-queues.md:
```diff
@@ -0,0 +1,22 @@
+
+In Versions below 6.2 DLQ will be enabled by default for TTBR messages as well.
+
+To opt-in to this behavior for Version 6.2 and higher use
```
Should be: Versions 6.2 and above, use:

https://github.com/particular/docs.particular.net/#version-language

andreasohlund replied:
Fixed

indualagarsamy commented on Snippets/Core/Core_6/Msmq/DLQOptInForTTBR.cs:
```diff
@@ -0,0 +1,17 @@
+            #region msmq-dlq-for-ttbr-optin
+
+            endpointConfiguration.UseTransport<MsmqTransport>()
+                .UseDeadLetterQueueForMessagesWithTimeToReachQueue();
```
The API says .TimeToReachQueue(), but isn't it TTBR? Since this is MSMQ specific, wouldn't it be better to stick to MSMQ terms?

andreasohlund replied:
The MSMQ term is `TimeToReachQueue` https://msdn.microsoft.com/en-us/library/ms711440(v=vs.85).aspx

indualagarsamy replied:
@andreasohlund - thanks. 
According to this article, there are 2 ways that messages can reach DLQ. When the TTBR or TTRQ timers elapse. This API talks about TTRQ. But the heartbeat etc, don't we set the TTBR?
https://blogs.msdn.microsoft.com/johnbreakwell/2010/01/22/where-have-my-msmq-messages-gone/

indualagarsamy replied:
For example: 
https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Transports/Msmq/MsmqMessageDispatcher.cs#L65-L67
This is all about not DLQing based on TTBR not TTRQ, yet the API is called TTRQ. IMO, that's incorrect. 
@SimonCropp - thoughts on this?

andreasohlund replied:
>According to this article, there are 2 ways that messages can reach DLQ. When the TTBR or TTRQ timers elapse. This API talks about TTRQ. But the heartbeat etc, don't we set the TTBR?

TTRQ is defaulted to infinite by us. Users can override this using the `timeToReachQueue` flag in the [connection string](https://docs.particular.net/nservicebus/msmq/connection-strings)

TTBR is the thing that users can control with `TimeToBeReceived` attribute. 

>This is all about not DLQing based on TTBR not TTRQ, yet the API is called TTRQ. IMO, that's incorrect.

I agree, will update the core API to `UseDeadLetterQueueForMessagesWithTimeToBeReceived` cc @Particular/nservicebus-maintainers

andreasohlund commented:
Core PR raised with new name https://github.com/Particular/NServiceBus/pull/4488 , this PR has been updated with the new name

weralabaj commented:
@andreasohlund seems you missed changes in the project config file (bump core nuget version)

andreasohlund commented:
Haven't bumped yet since the core PR isn't merged (will do once done)

indualagarsamy commented:
@andreasohlund - there are some content verification errors: 

```
System.Exception : The link /nservicebus/messaging/publish-subscribe/#mechanics-persistence-based doesn't exist, but was referenced in page \servicecontrol\contracts.md
The link /nservicebus/transports/transactions#transactions-transport-transaction doesn't exist, but was referenced in page \nservicebus\transports\transactions.md?version=Core_4

andreasohlund commented:
Hmm, I haven't changed any of those links?

On Tue, Feb 21, 2017 at 7:29 PM, Indu Alagarsamy <notifications@github.com>
wrote:

> @andreasohlund <https://github.com/andreasohlund> - there are some
> content verification errors:
>
> System.Exception : The link /nservicebus/messaging/publish-subscribe/#mechanics-persistence-based doesn't exist, but was referenced in page \servicecontrol\contracts.md
> The link /nservicebus/transports/transactions#transactions-transport-transaction doesn't exist, but was referenced in page \nservicebus\transports\transactions.md?version=Core_4
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Particular/docs.particular.net/pull/2437#issuecomment-281433295>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHoZAb08X0nE8ymhid6xXMHD0gmqWo3ks5reyzugaJpZM4L_nYH>
> .
>

SimonCropp commented:
U need to rebase

andreasohlund commented:
rebased

On Tue, Feb 21, 2017 at 7:46 PM, Simon Cropp <notifications@github.com>
wrote:

> U need to rebase
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Particular/docs.particular.net/pull/2437#issuecomment-281438339>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHoZD_YL5GBwESVFpx8-XITWeL52jBlks5rezEQgaJpZM4L_nYH>
> .
>

andreasohlund commented:
:+1: , updated according to your suggestion @indualagarsamy

On Mon, Feb 27, 2017 at 8:08 PM, Indu Alagarsamy <notifications@github.com>
wrote:

> *@indualagarsamy* commented on this pull request.
> ------------------------------
>
> In nservicebus/msmq/dead-letter-queues.md
> <https://github.com/Particular/docs.particular.net/pull/2437#discussion_r103285451>
> :
>
> > @@ -0,0 +1,22 @@
> +---
> +title: MSMQ Dead Letter Queues
> +summary: Controlling MSMQ Dead Letter Queue behavior
> +reviewed: 2017-02-13
> +tags:
> + - Transport
> +---
> +
> +[Dead Letter Queues](https://msdn.microsoft.com/en-us/library/ms706227.aspx) is a feature of MSMQ that tracks messages that are undeliverable, deleted, expired etc. NServiceBus endpoints will by default enable DLQ for all outgoing messages except messages that have a [Time To Be Received(TTBR)](/nservicebus/messaging/discard-old-messages.md) set. This avoids expired messages ending up in the DLQ and wasting disk space on the machine where they timed out.
>
> How about:
>
> @andreasohlund <https://github.com/andreasohlund> - Maybe we should a
> little more info. For example:
>
> Instead of
>
> This avoids expired messages ending up in the DLQ and wasting disk space
> on the machine where they timed out.
>
> how about:
>
> This avoids expired messages ending up either in the TDLQ or DLQ (if non-transactional
> queues are used <http:///nservicebus/msmq/connection-strings.md>) wasting
> disk space on the machine when they timed out.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Particular/docs.particular.net/pull/2437#discussion_r103285451>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHoZCnYkZPmF45IfVfwCfbPh3P7acbVks5rgx83gaJpZM4L_nYH>
> .
>

cremor commented on nservicebus/msmq/dead-letter-queues.md:
```diff
@@ -0,0 +1,22 @@
+
+### Enabling DLQ for messages with TTBR
+
+In Versions 6.2 and below DLQ will be enabled by default for TTBR messages as well.
```
Shouldn't this say "In versions below 6.2 ..."?

indualagarsamy replied:
@cremor - We use the following language convention when using versions in docs.
https://github.com/particular/docs.particular.net/#version-language

cremor replied:
@indualagarsamy English is not my first language, so maybe I'm wrong here. But if the first paragraph says

> In Versions 6.2 and below DLQ will be enabled by default

and the next paragraph says

> To opt-in to this behavior for Versions 6.2 and above, use

how should a user know if 6.2 does it by default or opt-in is required? Reading them as a programmer would read if-conditions, they are both true for version 6.2.

indualagarsamy replied:
@cremor - I can see your confusion. Good catch.   @andreasohlund - shouldn't this be Versions 6.1 and below that DLQ is enabled by default?

How about the below:

When using Versions 6.1 and below, DLQ will be enabled by default for messages that have the TTBR set.  However starting with Versions 6.2 and above, this is no longer enabled by default. To opt-in  for DLQ to be enabled for TTBR messages, use:

snippet: msmq-dlq-for-ttbr-optin

andreasohlund replied:
@indualagarsamy yea, that reads better, updated

@cremor thanks!!

andreasohlund commented:
Added upgrade guide entry https://github.com/Particular/docs.particular.net/pull/2437/commits/ba8ce9e3ed615e91b87d0e913db9323ea7763c9b
